### ISSUE TITLE
Unable to understand the encoder loss

### ISSUE THREAD

**Issue by nins-k** [houssamzenati/Efficient-GAN-Anomaly-Detection#8]
Hello,

I have a question regarding the loss calculation for the encoder.
`loss_encoder = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.zeros_like(l_encoder),logits=l_encoder))`

As per my understanding, the encoder is being optimized to have the discriminator output "0" (Fake) for the input combination of (x, E(X)) i.e. (Real Image, Encoder output of the Real Image). Could you please elaborate on why this is so or point me towards some relevant material?

I have read the associated research paper and the BiGAN paper which states that the generator and encoder should be updated together but I'm not able to relate it to how the encoder loss is calculated here.

Am I missing something obvious?

**Comment by houssamzenati** [comment]
Hi,

Apologies if it is not written in proper Latex:

So the min max game you consider in a BiGAN is : min_{E,G}max{D} V(E,G,D) (where E encoder, G generator and D discriminator).

We have V(E,G,D) = Expectation_{x real data} [logD(x,E(x))] +   Expectation_{z 'real' random seed} [log(1 -D(G(z),z))]

Therefore,

1) With regards to the discriminator, we want to maximize 
a) Expectation_{x real data} [logD(x,E(x))]  , thus D(x,E(x)) to be as close to 1 as possible. the pair x,E(x) is real for the Discriminator w.r.t to the Discriminator because x is real
        loss_dis_enc = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.ones_like(l_encoder),logits=l_encoder))
b) Expectation_{z 'real' random seed}[log(1 -D(G(z),z))], thus D(G(z),z)) to be as close to 0 as possible so that the whole argument inside the log is maximized. The pair G(z),z) is fake for the Discriminator w.r.t to the Discriminator because G(z) is fake

        loss_dis_gen = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.zeros_like(l_generator),logits=l_generator))
        loss_discriminator = loss_dis_gen + loss_dis_enc

2) With regards to the encoder, we want to minimize
a) Expectation_{x real data} [logD(x,E(x))]  , thus D(x,E(x)) to be as close to 0 as possible. the pair x,E(x) is fake for the Discriminator w.r.t to the Encoder/Generator because E(x) is not a "real" random seed.
        # encoder
loss_encoder = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.zeros_like(l_encoder),logits=l_encoder))
3) With regards to the generator, we want to minimize
b) Expectation_{z 'real' random seed}[log(1 -D(G(z),z))], thus D(G(z),z)) to be as close to 1 as possible so that the whole argument inside the log is minimized. The pair G(z),z) is real for the Discriminator w.r.t to the Encoder/Generator because z is a "real" random seed.
        # generator
        loss_generator = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.ones_like(l_generator),logits=l_generator))

I don't think it is obvious and it was actually a relevant issue. Hope it was useful. Please ask again if you still need some help.

Many thanks for your interest in our research,
Houssam

**Comment by nins-k** [comment]
Hi Houssam,

Thanks for your detailed response! 
It was very insightful and I believe I have a much better understanding of the loss function now.

Further, to form an intuitive reasoning of the encoder loss, I have the following line of though (I hope it's right):
    1. The Discriminator is trained to output "0" (Fake) for the input of (z, G(z)).
    2. The Encoder is optimized such that Discriminator returns 0 for (E(x), x).
    3. Hence, Encoder learns to create E(x) which is similar to z.

Thanks for your help and looking forward to reading your next paper (ALAD)  soon.
